// @stopify/stopify/dist/src/runtime/abstractRunner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const suspend_1 = require("./suspend");
const makeEstimator_1 = require("./makeEstimator");
const setImmediate_1 = require("./setImmediate");
var EventProcessingMode;
(function (EventProcessingMode) {
    EventProcessingMode[EventProcessingMode["Running"] = 0] = "Running";
    EventProcessingMode[EventProcessingMode["Paused"] = 1] = "Paused";
    EventProcessingMode[EventProcessingMode["Waiting"] = 2] = "Waiting";
})(EventProcessingMode = exports.EventProcessingMode || (exports.EventProcessingMode = {}));
class AbstractRunner {
    constructor(opts) {
        this.opts = opts;
        this.kind = 'ok';
        this.onDone = (result) => { };
        this.onYield = function () { };
        this.onBreakpoint = function () { };
        this.breakpoints = [];
        // The runtime system starts executing the main body of the program.
        this.eventMode = EventProcessingMode.Running;
        this.eventQueue = [];
        this.onYieldFlag = { kind: 'resume' };
        this.mayYieldFlag = { kind: 'resume' };
        // The global object for Stopified code.
        this.g = Object.create(null);
    }
    mayYieldRunning() {
        const n = this.suspendRTS.linenum;
        if (typeof n !== 'number') {
            return false;
        }
        return this.breakpoints.includes(n);
    }
    /**
     * Indirectly called by the stopified program.
     */
    init(rts) {
        this.continuationsRTS = rts;
        const estimator = makeEstimator_1.makeEstimator(this.opts);
        this.suspendRTS = new suspend_1.RuntimeWithSuspend(this.continuationsRTS, this.opts.yieldInterval, estimator, () => {
            switch (this.mayYieldFlag.kind) {
                case 'resume':
                    return this.mayYieldRunning();
                default: //Step
                    // Yield control if the line number changes.
                    const maybeLine = this.suspendRTS.linenum;
                    if (typeof maybeLine !== 'number' || maybeLine === this.mayYieldFlag.currentLine) {
                        return false;
                    }
                    else {
                        this.mayYieldFlag.onStep(maybeLine);
                        return true;
                    }
            }
        }, () => {
            if (this.eventMode === EventProcessingMode.Waiting) {
                throw new Error('Stopify internal error: onYield invoked during pause+wait');
            }
            switch (this.onYieldFlag.kind) {
                case 'paused':
                    this.onYieldFlag.onPaused(this.suspendRTS.linenum);
                    this.onYieldFlag = { kind: 'pausedAndMayYield' };
                    return false;
                case 'pausedAndMayYield':
                    throw new Error(`onYield called while paused. Do not call .resume()
            in the callback passed to .pause(). Resume in another turn.`);
                case 'resume':
                    if (this.mayYieldRunning()) {
                        this.onYieldFlag = {
                            kind: 'paused',
                            onPaused: (line) => {
                                if (line === undefined) {
                                    throw new Error(`Calling onBreakpoint with undefined as the line`);
                                }
                                this.onBreakpoint(line);
                            }
                        };
                        this.eventMode = EventProcessingMode.Paused;
                        // Invoke the breakpoint handler on the next turn, after the stack
                        // is fully reified.
                        setImmediate_1.setImmediate(() => this.onBreakpoint(this.suspendRTS.linenum));
                        return false;
                    }
                    this.onYield();
                    return true;
                default: // Step
                    return !this.suspendRTS.mayYield();
            }
        });
        return this;
    }
    /**
     * Called by the stopified program.
     */
    suspend() {
        return this.suspendRTS.suspend();
    }
    /**
     * Called by the stopified program.
     */
    onEnd(result) {
        this.eventMode = EventProcessingMode.Waiting;
        this.onDone(result);
        this.processQueuedEvents();
    }
    runInit(onDone, onYield, onBreakpoint) {
        if (onYield) {
            this.onYield = onYield;
        }
        if (onBreakpoint) {
            this.onBreakpoint = onBreakpoint;
        }
        this.onDone = onDone;
    }
    pause(onPaused) {
        if (this.eventMode === EventProcessingMode.Paused) {
            throw new Error('the program is already paused');
        }
        if (this.eventMode === EventProcessingMode.Waiting) {
            onPaused(); // onYield will not be invoked
        }
        else {
            this.onYieldFlag = { kind: 'paused', onPaused };
        }
        this.eventMode = EventProcessingMode.Paused;
    }
    setBreakpoints(lines) {
        this.breakpoints = lines;
    }
    resume() {
        if (this.eventMode === EventProcessingMode.Waiting) {
            return;
        }
        if (this.eventMode === EventProcessingMode.Running) {
            throw new Error(`invokes .resume() while the program is running`);
        }
        // Program was paused, but there was no active continuation.
        if (this.suspendRTS.continuation === suspend_1.badResume) {
            this.eventMode = EventProcessingMode.Waiting;
            this.processQueuedEvents();
        }
        else {
            this.eventMode = EventProcessingMode.Running;
            this.mayYieldFlag = { kind: 'resume' };
            this.onYieldFlag = { kind: 'resume' };
            this.suspendRTS.resumeFromCaptured();
        }
    }
    // NOTE: The program remains Paused while stepping.
    step(onStep) {
        if (this.eventMode !== EventProcessingMode.Paused) {
            throw new Error(`step(onStep) requires the program to be paused`);
        }
        this.mayYieldFlag = { kind: 'step', currentLine: this.suspendRTS.linenum, onStep };
        this.onYieldFlag = { kind: 'step' };
        this.suspendRTS.resumeFromCaptured();
    }
    // NOTE: In both of the pause functions below, we don't switch the mode to
    // paused! This is because we don't want the user to be able to hit "step" at
    // this point.
    pauseK(callback) {
        return this.continuationsRTS.captureCC((k) => {
            return this.continuationsRTS.endTurn(onDone => {
                return callback((x) => {
                    return this.continuationsRTS.runtime(() => k(x), onDone);
                });
            });
        });
    }
    pauseImmediate(callback) {
        return this.continuationsRTS.captureCC((k) => {
            return this.continuationsRTS.endTurn(onDone => {
                this.k = { k, onDone };
                callback();
            });
        });
    }
    continueImmediate(x) {
        if (this.k === undefined) {
            throw new Error(`called continueImmediate before pauseImmediate`);
        }
        const { k, onDone } = this.k;
        this.k = undefined;
        return this.continuationsRTS.runtime(() => k(x), onDone);
    }
    externalHOF(body) {
        return this.continuationsRTS.captureCC((k) => this.continuationsRTS.endTurn(onDone => body(result => this.continuationsRTS.runtime(() => k(result), onDone))));
    }
    runStopifiedCode(body, callback) {
        this.continuationsRTS.runtime(body, callback);
    }
    processEvent(body, receiver) {
        this.eventQueue.push({ body, receiver });
        this.processQueuedEvents();
    }
    processQueuedEvents() {
        if (this.eventMode !== EventProcessingMode.Waiting) {
            return;
        }
        const eventHandler = this.eventQueue.shift();
        if (eventHandler === undefined) {
            return;
        }
        const { body, receiver } = eventHandler;
        this.eventMode = EventProcessingMode.Running;
        this.continuationsRTS.runtime(body, (result) => {
            this.eventMode = EventProcessingMode.Waiting;
            receiver(result);
            this.processQueuedEvents();
        });
    }
}
exports.AbstractRunner = AbstractRunner;
